modules/packages/Sort.chpl

```diff
@@ -968,123 +968,196 @@ module MergeSort {
-      Data(hi) <=> Data(lo);
-    if (chpl_compare(Data(hi), Data(mid), comparator) < 0) then
-      Data(hi) <=> Data(mid);
+    // Now put piv back in Data[lo]
+    ShallowCopy.shallowCopy(Data[lo], piv); // leaves piv empty
 
-    const pivotVal = Data(mid);
-    Data(mid) = Data(hi-stride);
-    Data(hi-stride) = pivotVal;
-    // end median-of-3 partitioning
+    // now we are in the state:
+    //   |  =  |  <  |  > |  =   |
+    //    lo    a   c b  d     hi
 
-    var loptr = lo,
-        hiptr = hi-stride;
-    while (loptr < hiptr) {
-      do { loptr += stride; } while (chpl_compare(Data(loptr), pivotVal, comparator) < 0);
-      do { hiptr -= stride; } while (chpl_compare(pivotVal, Data(hiptr), comparator) < 0);
-      if (loptr < hiptr) {
-        Data(loptr) <=> Data(hiptr);
-      }
+    // now place the equal regions in the right places
+    var s, l, h: int;
+
+    // Fix the first = region
+    s = min(a-lo, b-a); // the number of of elements to swap
+    l = lo;
+    h = b-s;
+    while s > 0 {
+      ShallowCopy.shallowSwap(Data[l], Data[h]);
+      l += 1;
+      h += 1;
+      s -= 1;
+    }
+
+    // Fix the second = region
+    var n = hi+1;
+    s = min(d-c, hi-d);
+    l = b;
+    h = n-s;
+    while s > 0 {
+      ShallowCopy.shallowSwap(Data[l], Data[h]);
+      l += 1;
+      h += 1;
+      s -= 1;
     }
 
-    Data(hi-stride) = Data(loptr);
-    Data(loptr) = pivotVal;
+    var eqStart = b-a+lo;
+    var eqEnd = hi-(d-c);
 
-    // TODO -- Get this cobegin working and tested
-    //  cobegin {
-      quickSort(Data[..loptr-stride by stride align lo], minlen, comparator);  // could use unbounded ranges here
-      quickSort(Data[loptr+stride.. by stride align lo], minlen, comparator);
-    //  }
+    return (eqStart, eqEnd);
   }
 
-  /* Non-stridable quickSort */
-  proc quickSort(Data: [?Dom] ?eltType, minlen=16, comparator:?rec=defaultComparator)
-    where !Dom.stridable {
+
+  // Puts lo, mid, and hi in order by swapping.
+  proc order3(Data: [?Dom] ?eltType,
+              lo: int, mid: int, hi: int,
+              comparator) {
+    if (chpl_compare(Data[mid], Data[lo], comparator) < 0) then
+      ShallowCopy.shallowSwap(Data[mid], Data[lo]);
+    if (chpl_compare(Data[hi], Data[lo], comparator) < 0) then
+      ShallowCopy.shallowSwap(Data[hi], Data[lo]);
+    if (chpl_compare(Data[hi], Data[mid], comparator) < 0) then
+      ShallowCopy.shallowSwap(Data[hi], Data[mid]);
+  }
+
+  /* Non-stridable quickSort to sort Data[start..end] */
+  proc quickSort(Data: [?Dom] ?eltType,
+                 minlen=16,
+                 comparator:?rec=defaultComparator,
+                 start:int = Dom.low, end:int = Dom.high) {
 
     chpl_check_comparator(comparator, eltType);
 
     if Dom.rank != 1 {
       compilerError("quickSort() requires 1-D array");
     }
 
+    if Dom.stridable {
+      ref reindexed = Data.reindex(Dom.alignedLow..#Dom.size);
+      quickSort(reindexed, minlen, comparator);
```

Comment:
Doesn't this need to pass the reindexed `start` and `end`, in case they weren't the default values?

Reply:
I don't think so, because quickSort has default values for start/end based on the domain, so that will end up with Dom.alignedLow as the start e.g.

Reply:
But the caller may have passed non-default values, to sort only part of the array, no?

Reply:
Ah, I see what you are saying now.

Reply:
I made the stridable handling code not be in the version accepting start and end.